File: 프로그래머스/3/43164. 여행경로/여행경로.java
import java.util.*;

class Solution {
    boolean[] visited;
    List<String> answer;

    public String[] solution(String[][] tickets) {
        visited = new boolean[tickets.length];

        // 사전순 정렬
        Arrays.sort(tickets, (a, b) -> {
            if (a[0].equals(b[0])) return a[1].compareTo(b[1]);
            return a[0].compareTo(b[0]);
        });

        List<String> path = new ArrayList<>();
        path.add("ICN");

        dfs("ICN", tickets, path);

        return answer.toArray(new String[0]);
    }

    private boolean dfs(String cur, String[][] tickets, List<String> path) {
        if (path.size() == tickets.length + 1) {
            answer = new ArrayList<>(path);
            return true;
        }

        for (int i = 0; i < tickets.length; i++) {
            if (!visited[i] && tickets[i][0].equals(cur)) {
                visited[i] = true;
                path.add(tickets[i][1]);

                if (dfs(tickets[i][1], tickets, path)) return true;

                // 백트래킹
                visited[i] = false;
                path.remove(path.size() - 1);
            }
        }

        return false;
    }
}

File: 프로그래머스/3/43164. 여행경로/여행경로.js
function solution(tickets) {
    const visited = Array(tickets.length).fill(false);
    let answer = [];

    // 사전순 정렬
    tickets.sort((a, b) => {
        if (a[0] === b[0]) return a[1].localeCompare(b[1]);
        return a[0].localeCompare(b[0]);
    });

    function dfs(cur, path) {
        if (path.length === tickets.length + 1) {
            answer = [...path];
            return true;
        }

        for (let i = 0; i < tickets.length; i++) {
            if (!visited[i] && tickets[i][0] === cur) {
                visited[i] = true;
                path.push(tickets[i][1]);

                if (dfs(tickets[i][1], path)) return true;

                visited[i] = false;
                path.pop();
            }
        }

        return false;
    }

    dfs("ICN", ["ICN"]);
    return answer;
}